lib/3scale/backend/validators/key.rb

```diff
@@ -3,12 +3,18 @@ module Backend
     module Validators
       class Key < Base
         def apply
-          if service.backend_version.to_i == 1 ||
-              application.has_no_keys? ||
-              application.has_key?(params[:app_key])
-            succeed!
-          else
+          case service.backend_version.to_i
+          when 1 # Auth by user_key
+            return succeed! if params.has_key?(:user_key)
+
+            fail!(UserKeyInvalid.new(params[:user_key]))
+          when 2, 0  # 2 -> Auth by app_id. 0 -> not defined or invalid, must behave like 2
```

Comment:
hm, is it actually possible to have `backend_version.to_i == 0` ? In what scenario?
Not sure if it's right...

Reply:
when `backend_version` is not a valid number, `to_i` returns `0`. AFAIK `backend_version` can only be `1`, `2` and `oauth`. So in this case it will be `0` for `oauth`.